### ISSUE TITLE
Prepared statement inserts don't work with materialized columns 

### ISSUE THREAD

**Issue by sholokhov** [ClickHouse/clickhouse-jdbc#877]
Hi! 

After switching from `0.3.1` driver we started to see that inserts are failing if table has a materialized columns. We prepare insert as following:
```
insert into distr.table_test (session_id, event, event_type) values (?, ?, ?)
```
where distr.table_test is a Distributed table with additionally defined `mat_rating` column. On `0.3.1` insert works fine, but on `0.3.2-patch{6,7}` (I've tested those 2) it fails with:

> Code: 44. DB::Exception: Cannot insert column mat_rating, because it is MATERIALIZED column. (ILLEGAL_COLUMN) (version 22.2.2.1)\n, server ClickHouseNode(addr=http:10.132.0.211/<unresolved>:8123, db=default)@-1441372095\n\tat com.clickhouse.jdbc.SqlExceptionUtils.batchUpdateError(SqlExceptionUtils.java:90)\n\tat com.clickhouse.jdbc.internal.InputBasedPreparedStatement.executeAny(InputBasedPreparedStatement.java:130)\n\tat com.clickhouse.jdbc.internal.AbstractPreparedStatement.executeLargeBatch(AbstractPreparedStatement.java:85)\n\tat com.clickhouse.jdbc.internal.ClickHouseStatementImpl.executeBatch(ClickHouseStatementImpl.java:539)\n\tat com.zaxxer.hikari.pool.ProxyStatement.executeBatch(ProxyStatement.java:127)\n\tat com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeBatch(HikariProxyPreparedStatement.java)\n\tat doobie.free.KleisliInterpreter$PreparedStatementInterpreter.$anonfun$executeBatch$2(kleisliinterpreter.scala:867)\n\tat doobie.free.KleisliInterpreter.$anonfun$primitive$2(kleisliinterpreter.scala:113)\n\tat blocking @ doobie.util.transactor$Transactor$fromDataSource$FromDataSourceUnapplied.release$1(transactor.scala:308)\n\tat flatMap @ doobie.WeakAsync$$anon$1.flatMap(WeakAsync.scala:28)\n\tat flatMap @ doobie.WeakAsync$$anon$1.flatMap(WeakAsync.scala:28)\n\tat blocking @ doobie.util.transactor$Transactor$fromDataSource$FromDataSourceUnapplied.release$1(transactor.scala:308)\n\tat flatMap @ doobie.WeakAsync$$anon$1.flatMap(WeakAsync.scala:28)\n\tat flatMap @ doobie.WeakAsync$$anon$1.flatMap(WeakAsync.scala:28)\n\tat blocking @ doobie.util.transactor$Transactor$fromDataSource$FromDataSourceUnapplied.release$1(transactor.scala:308)\n\tat flatMap @ doobie.WeakAsync$$anon$1.flatMap(WeakAsync.scala:28)\n\tat flatMap @ doobie.WeakAsync$$anon$1.flatMap(WeakAsync.scala:28)\n\tat blocking @ doobie.util.transactor$Transactor$fromDataSource$FromDataSourceUnapplied.release$1(transactor.scala:308)

Seems to be a regression?

**Comment by zhicwu** [comment]
> Seems to be a regression?

Could be. What's version of ClickHouse? And the table DDL? I need them to reproduce the issue first.

**Comment by sholokhov** [comment]
Clickhouse version is 22.2.2.1.
SQL:

```sql
create table reporting_be_local
(
    timestamp            DateTime64(3, 'UTC'),
    event                    LowCardinality(String),
    event_type           LowCardinality(String),
    session_id           UInt64,
    mat_rating           UInt8 materialized multiIf(event = 'search', 1, event = 'channel', 2, 0) -- simplified 
)
    engine = ReplicatedReplacingMergeTree('/clickhouse/tables/{cluster}/{shard}/reporting_be_local', '{replica}')
        PARTITION BY toYYYYMMDD(timestamp)
        ORDER BY (timestamp, session_id, event)
        SETTINGS index_granularity = 8192;

create table reporting_be
(
    timestamp            DateTime64(3, 'UTC'),
    event                LowCardinality(String),
    event_type           LowCardinality(String),
    session_id           UInt64,
    mat_rating           UInt8
)
engine = Distributed('{cluster}', 'shard', 'reporting_be_local', murmurHash3_64(session_id));
```

**Comment by zhicwu** [comment]
Thanks @sholokhov. Why `mat_rating UInt8` is needed for the distributed table? I can reproduce the issue even with CLI client. Removing the column seems working.

**Comment by sholokhov** [comment]
We actually run read queries against this distributed table and use it to distribute data during inserts. So we need this field to be presented in distributed table. Or is there any workourd for this?

**Comment by zhicwu** [comment]
@sholokhov, I don't think it's an issue with any ClickHouse client but the server. Just did a quick search, a similar case were first reported in ClickHouse/ClickHouse#2331, and maybe someone can take [this](https://github.com/ClickHouse/ClickHouse/blob/2e22164cd7e97bba14abedca8dcbf472e008db3a/tests/integration/test_sharding_key_from_default_column/test.py#L56-L77) integration test as an example to cover your case.

Anyway, I assume you don't like the idea of using two distributed tables(one for query and the other for insert). So you may want to change your distributed table by changing `mat_rating UInt8` to something like `mat_rating UInt8 MATERIALIZED -1` and that's it. FYI, I also tried to change it to nullable column and/or insert with settings but none of them worked.

Lastly, I hope you don't mind that I took above DDL and added an integration test case in JDBC driver.

**Comment by sholokhov** [comment]
Hmm, interesting! I've tried with `mat_rating UInt8 MATERIALIZED -1` and it worked fine, thanks! Will also play around with splitting the distributed table on RO/RW, but indeed it seems to be less convinient because of other dependencies. Took a look on the issue you've mentioned and curios how it was solved for them? 🤔

Anyways, thanks for help! I don't mind if you use provided DLL in integration test, happy if it helps! :)
